accepted/2020/serializer/WriteableDomAndDynamic.md

```diff
@@ -0,0 +1,587 @@
+```cs
+    JsonNode node = ... // a non-dynamic JsonNode
+    dynamic obj = node.ToDynamic();
+```
+
+The implementation of `ToDynamic()` is efficient since it will perform a shallow clone by copying the internal value (for `JsonValue`), array (for `JsonArray`) or dictionary (for `JsonObject`):
+```cs
+    JsonNode node = ... // a non-dynamic JsonNode
+    byte[] utf8 = node.SerializeToUtf8Bytes();
+    dynamic obj = JsonSerializer.Deserialize<JsonNode>(utf8, options);
+```
+
+# Writable DOM
+A deserialized DOM is based upon JSON token types (object, array, number, string, true, false) since the corresponding CLR types are not known, unlike deserializing against a strongly-typed POCO. This is consistent with `JsonElement` semantics as well.
+
+Although a deserializated DOM in "read mode" fully maps to the JSON token types, once in "edit mode", CLR values to be assigned to nodes which do not necessarily map to JSON token types. Two common examples:
+- Quoted numbers. A CLR number (`Double`, `Int32`, etc) is normally serialized to JSON as a number. However, a number may be serialized as a JSON string instead - which is necessary to support CLR numbers that can't be represented in JSON as a number, such as `Double.NaN` or `Double.PositiveInfinity`.
+- String Enums. A CLR `Enum` is normally serialized to JSON as a number, but through a custom converter (including `JsonStringEnumConverter`), an `Enum` can be serialized as the string representation.
+
+For example, consider this **hypothetical API** that has a `JsonNumber` class (and not the proposal of using just `JsonNode`)
+```cs
+// Ask for a number with a JSON number token type.
+// Internally probably stored as a byte[] or JsonElement.
+JsonNumber number = JsonSerializer.Deserialize<JsonNumber>("1.2");
+double dlb = number.GetValue<double>();
+
+// Change to "NaN".
+number.Value = double.Nan;
+string json = number.Serialize();
+Debug.Assert(json == "\"NaN\""); // Due to quoted number support, this works.
+
+// However what about round-tripping?
+// Can a JSON string be deserialized into a JsonNumber? If so, "JsonNumber" no longer means a "JSON number".
+// When deserializing a string there is no known CLR type, so the internal value would need be based on a string\byte[].
+// So a "JsonString" node would normally be created, not "JsonNumber" and thus throw here:
+number = JsonSerializer.Deserialize<JsonNumber>(json); // Throws
+
+// Even if the above could be made to work, it wouldn't work with a JsonObject node.
+// A JsonObject would create "JsonString" nodes for "NaN" JSON strings since it has
+// no way of knowing these should map to a JsonNumber (or a CLR number).
+```
+
+Thus the DOM is based on CLR types (not JSON tokens) in edit mode and during serialization. The JSON "value" types (number, string, boolean) are represented as a common `JsonValue` sealed class, and **not** separate `JsonNumber`, `JsonString` and `JsonBoolean` classes as shown in the example above.
```

Comment:
We should revisit this design choice. Are there substantial customer benefits of treating the DOM as CLR types (which we can't solve in other ways), that is worth potentially subverting user expectations?

If we really want to support string to number conversion, another way to do so would be via an specific options on the parse/ctor of the JsonNode/JsonNumber types.